File: examples/03-raw-transactions/typescript/main.ts
/**
 * Lab 03: Raw Transactions (TypeScript Reference Implementation)
 * Demonstrates manual UTXO selection, fee calculation, raw transaction construction,
 * signing, mempool acceptance testing, and broadcast.
 */
import { bootstrapLab } from "../../typescript/src/common/bootstrap.js";

interface UTXO {
  txid: string;
  vout: number;
  amount: number;
}

interface SignResult {
  hex: string;
  complete: boolean;
}

interface MempoolAcceptResult {
  txid: string;
  allowed: boolean;
  rejectReason?: string;
}

async function main() {
  console.log("=== Lab 03: Raw Transactions (TypeScript) ===");

  try {
    // Step 1: Bootstrap lab
    process.stdout.write("[Step 1] Bootstrapping lab wallet & ensure spendable UTXOs ... ");
    const { rpc: nodeRpc, walletRpc } = await bootstrapLab();
    console.log("✓");

    // Step 2: Select UTXO from listunspent
    process.stdout.write("[Step 2] Selecting UTXO from listunspent ... ");
    const unspent = await walletRpc.call<UTXO[]>("listunspent", [1, 9999999]);
    if (!unspent || unspent.length === 0) {
      throw new Error("No spendable UTXOs found in wallet");
    }

    const utxo = unspent[0];
    console.log(`✓ (txid: ${utxo.txid.substring(0, 16)}..., vout: ${utxo.vout}, amount: ${utxo.amount} BTC)`);

    // Step 3: Construct raw transaction
    process.stdout.write("[Step 3] Constructing raw transaction hex (createrawtransaction) ... ");
    const recipientAddr = await walletRpc.call<string>("getnewaddress", ["recipient", "bech32m"]);
    const changeAddr = await walletRpc.call<string>("getnewaddress", ["change", "bech32m"]);

    const sendAmount = 1.5;
    const fee = 0.0001;
    const changeAmount = Number((utxo.amount - sendAmount - fee).toFixed(8));
    if (changeAmount <= 0) {
      throw new Error(`UTXO amount (${utxo.amount}) insufficient for send + fee`);
    }

    const inputs = [{ txid: utxo.txid, vout: utxo.vout }];
    const outputs = [
      { [recipientAddr]: sendAmount },
      { [changeAddr]: changeAmount },
    ];

    const rawTxHex = await walletRpc.call<string>("createrawtransaction", [inputs, outputs]);
    if (!rawTxHex) throw new Error("createrawtransaction failed");
    console.log(`✓ (hex length: ${rawTxHex.length})`);

    // Step 4: Sign transaction inputs
    process.stdout.write("[Step 4] Signing transaction inputs (signrawtransactionwithwallet) ... ");
    const signResult = await walletRpc.call<SignResult>("signrawtransactionwithwallet", [rawTxHex]);
    if (!signResult.complete) {
      throw new Error("Transaction signing was incomplete");
    }
    console.log("✓ (complete: true)");

    // Step 5: Verify transaction with testmempoolaccept
    process.stdout.write("[Step 5] Verifying transaction with testmempoolaccept ... ");
    const mempoolTest = await nodeRpc.call<MempoolAcceptResult[]>("testmempoolaccept", [[signResult.hex]]);
    if (!mempoolTest || !mempoolTest[0]?.allowed) {
      throw new Error(`Mempool reject: ${JSON.stringify(mempoolTest)}`);
    }
    console.log("✓ (allowed: true)");

    // Step 6: Broadcast transaction via sendrawtransaction
    process.stdout.write("[Step 6] Broadcasting transaction via sendrawtransaction ... ");
    const broadcastTxid = await nodeRpc.call<string>("sendrawtransaction", [signResult.hex]);
    if (!broadcastTxid || broadcastTxid.length !== 64) {
      throw new Error(`Invalid broadcast txid: ${broadcastTxid}`);
    }
    console.log(`✓ (txid: ${broadcastTxid})`);

    // Step 7: Mine 1 block & verify confirmation
    process.stdout.write("[Step 7] Mining 1 block & verifying confirmation ... ");
    const minerAddr = await walletRpc.call<string>("getnewaddress", ["miner", "bech32m"]);
    await walletRpc.call("generatetoaddress", [1, minerAddr]);

    const txDetails = await walletRpc.call<{ confirmations: number }>("getrawtransaction", [broadcastTxid, true]);
    if (!txDetails.confirmations || txDetails.confirmations < 1) {
      throw new Error(`Expected confirmations >= 1, got ${txDetails.confirmations}`);
    }
    console.log(`✓ (confirmations: ${txDetails.confirmations})`);

    console.log("======================================================");
    console.log("Result: PASS ✓");
    console.log("======================================================");
    process.exit(0);
  } catch (err: any) {
    console.error(`\n✗ FAILURE: ${err.message || err}`);
    console.log("======================================================");
    console.log("Result: FAIL ✗");
    console.log("======================================================");
    process.exit(1);
  }
}

main();

File: examples/03-raw-transactions/python/main.py
#!/usr/bin/env python3
"""
Lab 03: Raw Transactions (Python Reference Implementation)
Demonstrates manual UTXO selection, fee calculation, raw transaction construction,
signing, mempool acceptance testing, and broadcast.
"""
import sys
from decimal import Decimal, ROUND_DOWN
from pathlib import Path

# Ensure examples/python is in sys.path
SCRIPT_DIR = Path(__file__).resolve().parent
REPO_ROOT = SCRIPT_DIR.parent.parent
PYTHON_ROOT = REPO_ROOT / "examples" / "python"
if str(PYTHON_ROOT) not in sys.path:
    sys.path.insert(0, str(PYTHON_ROOT))

from common.rpc import BitcoinRPC
from common.bootstrap import bootstrap_lab


def main():
    print("=== Lab 03: Raw Transactions (Python) ===")

    try:
        # Step 1: Bootstrap lab
        print("[Step 1] Bootstrapping lab wallet & ensure spendable UTXOs ... ", end="", flush=True)
        node_rpc, wallet_rpc = bootstrap_lab()
        print("✓")

        # Step 2: Select UTXO from listunspent
        print("[Step 2] Selecting UTXO from listunspent ... ", end="", flush=True)
        unspent = wallet_rpc.call("listunspent", [1, 9999999])
        if not unspent:
            raise RuntimeError("No spendable UTXOs found in wallet")

        # Pick the largest UTXO — the shared `lab` wallet also holds small change
        # outputs from earlier lab runs, so the first entry is not reliably big enough.
        utxo = max(unspent, key=lambda u: u["amount"])
        utxo_txid = utxo["txid"]
        utxo_vout = utxo["vout"]
        utxo_amount = Decimal(str(utxo["amount"]))
        print(f"✓ (txid: {utxo_txid[:16]}..., vout: {utxo_vout}, amount: {utxo_amount} BTC)")

        # Step 3: Construct raw transaction
        print("[Step 3] Constructing raw transaction hex (createrawtransaction) ... ", end="", flush=True)
        recipient_addr = wallet_rpc.call("getnewaddress", ["recipient", "bech32m"])
        change_addr = wallet_rpc.call("getnewaddress", ["change", "bech32m"])

        send_amount = Decimal("1.5")
        fee = Decimal("0.0001")
        change_amount = utxo_amount - send_amount - fee
        if change_amount <= 0:
            raise RuntimeError(f"UTXO amount ({utxo_amount}) too small for send + fee")

        inputs = [{"txid": utxo_txid, "vout": utxo_vout}]
        outputs = [
            {recipient_addr: float(send_amount)},
            {change_addr: float(change_amount)},
        ]

        raw_tx_hex = wallet_rpc.call("createrawtransaction", [inputs, outputs])
        assert isinstance(raw_tx_hex, str) and len(raw_tx_hex) > 0, "Invalid raw transaction hex"
        print(f"✓ (hex length: {len(raw_tx_hex)})")

        # Step 4: Sign transaction inputs
        print("[Step 4] Signing transaction inputs (signrawtransactionwithwallet) ... ", end="", flush=True)
        sign_result = wallet_rpc.call("signrawtransactionwithwallet", [raw_tx_hex])
        assert sign_result.get("complete") is True, "Signing incomplete"
        signed_hex = sign_result["hex"]
        print("✓ (complete: true)")

        # Step 5: Verify transaction with testmempoolaccept
        print("[Step 5] Verifying transaction with testmempoolaccept ... ", end="", flush=True)
        mempool_test = node_rpc.call("testmempoolaccept", [[signed_hex]])
        assert mempool_test and mempool_test[0].get("allowed") is True, f"Mempool reject: {mempool_test}"
        print("✓ (allowed: true)")

        # Step 6: Broadcast transaction via sendrawtransaction
        print("[Step 6] Broadcasting transaction via sendrawtransaction ... ", end="", flush=True)
        broadcast_txid = node_rpc.call("sendrawtransaction", [signed_hex])
        assert isinstance(broadcast_txid, str) and len(broadcast_txid) == 64, "Invalid broadcast txid"
        print(f"✓ (txid: {broadcast_txid})")

        # Step 7: Mine 1 block and verify confirmation
        print("[Step 7] Mining 1 block & verifying confirmation ... ", end="", flush=True)
        miner_addr = wallet_rpc.call("getnewaddress", ["miner", "bech32m"])
        wallet_rpc.call("generatetoaddress", [1, miner_addr])

        tx_details = wallet_rpc.call("getrawtransaction", [broadcast_txid, True])
        confs = tx_details.get("confirmations", 0)
        assert confs >= 1, f"Expected confirmations >= 1, got {confs}"
        print(f"✓ (confirmations: {confs})")

        print("======================================================")
        print("Result: PASS ✓")
        print("======================================================")
        sys.exit(0)

    except Exception as exc:
        print(f"\n✗ FAILURE: {exc}")
        print("======================================================")
        print("Result: FAIL ✗")
        print("======================================================")
        sys.exit(1)


if __name__ == "__main__":
    main()
